sqlx-macros/src/query/mod.rs

```diff
@@ -157,67 +158,119 @@ pub fn expand_input(input: QueryMacroInput) -> crate::Result<TokenStream> {
     }
 }
 
-#[allow(unused_variables)]
-fn expand_from_db(input: QueryMacroInput, db_url: &str) -> crate::Result<TokenStream> {
-    // FIXME: Introduce [sqlx::any::AnyConnection] and [sqlx::any::AnyDatabase] to support
-    //        runtime determinism here
-
-    let db_url = Url::parse(db_url)?;
-    match db_url.scheme() {
-        #[cfg(feature = "postgres")]
-        "postgres" | "postgresql" => {
-            let data = block_on(async {
-                let mut conn = sqlx_core::postgres::PgConnection::connect(db_url.as_str()).await?;
-                QueryData::from_db(&mut conn, &input.sql).await
-            })?;
-
-            expand_with_data(input, data, false)
-        },
-
-        #[cfg(not(feature = "postgres"))]
-        "postgres" | "postgresql" => Err("database URL has the scheme of a PostgreSQL database but the `postgres` feature is not enabled".into()),
-
-        #[cfg(feature = "mssql")]
-        "mssql" | "sqlserver" => {
-            let data = block_on(async {
-                let mut conn = sqlx_core::mssql::MssqlConnection::connect(db_url.as_str()).await?;
-                QueryData::from_db(&mut conn, &input.sql).await
-            })?;
-
-            expand_with_data(input, data, false)
-        },
-
-        #[cfg(not(feature = "mssql"))]
-        "mssql" | "sqlserver" => Err("database URL has the scheme of a MSSQL database but the `mssql` feature is not enabled".into()),
+#[derive(Debug)]
+enum ConnectionCacheItem {
+    #[cfg(feature = "postgres")]
+    Postgres(sqlx_core::postgres::PgConnection),
+    #[cfg(feature = "mssql")]
+    MsSql(sqlx_core::mssql::MssqlConnection),
+    #[cfg(feature = "mysql")]
+    MySql(sqlx_core::mysql::MySqlConnection),
+    #[cfg(feature = "sqlite")]
+    Sqlite(sqlx_core::sqlite::SqliteConnection),
+}
```

Comment:
This is so annoyingly close to `AnyConnection` that I really hate seeing it defined twice.

The main issue with using `AnyConnection` here is that you'd want to destructure it to get into the database-specific implementations, but that's not exposed. I think I'd be okay exposing the minimum parts necessary (`AnyConnection::kind()`, `AnyConnectionKind`) as long as they're `#[doc(hidden)]`, though. I did consider suggesting an `unstable` feature here but `sqlx-macros` would have to enable that feature by default which makes it rather pointless anyway.

Reply:
I agree, using `AnyConnection` would certainly simplify things. I'll work on switching things over tomorrow if I have time

Reply:
8905a258f604defe4c9909c5119a12ebcd1221bf switches things over to using an `AnyConnection` by adding a method that exposes a mutable reference to the inner `AnyConnectionKind` value. This does cause `sqlx-macros` to have a confusing compiler error when trying to build without any database features because `any` in `sqlx-core` requires the `any` feature, and at least one of the database features to be enabled